### ISSUE TITLE
Longformer, BigBird take same time to run in sparse mode as well as full-mode

### ISSUE THREAD

**Issue by allohvk** [huggingface/transformers#18234]
### System Info

Transformers: 4.20.1
Python: 3.8.12
Pretrained models & tokenizer from HF: "allenai/longformer-base-4096" and "google/bigbird-roberta-base"

Longformer: Take same time to train (finetume) a pretrained model for different sliding window sizes of 256, 512, 1024 or 2048. One would expect that at lower sliding window sizes, the training times should be lower.

BigBird: Same problem as above. In fact BigBird has a simple switch to change from sparse-attention to full-attention. The training time taken in both cases is roughly the same which seems to point to some issue.

Small but complete source code to simulate: 
https://colab.research.google.com/drive/1nm7a-qJseNSCkAB5_3QNkVSrHc8zePAV?usp=sharing

### Who can help?

@ydshieh 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

https://colab.research.google.com/drive/1nm7a-qJseNSCkAB5_3QNkVSrHc8zePAV?usp=sharing

### Expected behavior

Longformer: Take different time to train (finetume) a pretrained model for different sliding window sizes of 256, 512, 1024 or 2048. One would expect that at lower sliding window sizes, the training times should be lower.

BigBird: Same problem as above. In fact BigBird has a simple switch to change from sparse-attention to full-attention. The training time taken in both cases is roughly the same which seems to point to some issue.

**Comment by allohvk** [comment]
@ydshieh It gets a bit more wierder. Today I tried to directly use the Longformer bypassing Huggingface. 

It needed minor changes to the above code. The link is here: 
https://colab.research.google.com/drive/1R5uDsbl3ZmUIccZtefVNBs3CXU_vcDZd?usp=sharing

The observations continues to be perplexing:
CASE 1:
ATT_MODE = 'sliding_chunks'; 100% LOCAL attention ie ATT_TYPE = 1 for all tokens
SLIDE_WIN_SIZE = 256(default) takes between 9-10 hours to train
SLIDE_WIN_SIZE = 1024 takes between 9-10 hours to train
Observation: Sparse attention with 256 tokens windowsize should not take same fine-tuning time as 1024 tokens

CASE 2:
ATT_MODE = 'sliding_chunks'; NO attention ie ATT_TYPE = 0 for all tokens
SLIDE_WIN_SIZE is immaterial
Observation: It is observed that even if none of tokens attend to each other, training time taken is same as case 1 
above ie 9-10 hours which should not be the case

CASE 3:
ATT_MODE = 'sliding_chunks'; 100% Global attention: : ATT_TYPE = 2
SLIDE_WIN_SIZE is immaterial
Observation: With 100% global attention, every token attends to each other. It is observed that if all tokens attend to each other, training time taken is 16-17 hours. This training time should be similar to Case 4 which is NOT the case

Case 4: This is the most bizzarre
ATT_MODE = 'n2'
We can simply set choose the attention mode = 'n2' which is regular quadratic attention. Theoritically this should take
same training time as Case 3 (when all tokens are marked as global)
Observation: n2 attention takes the lowest training time of approx 2 hours only which is exact opposite of what LOngformer is supposed to do !!!

Should I open a bug directly with the Longformer GITHUB?

**Comment by ydshieh** [comment]
| First Header  | Tiny | Base | Large |
| ------------- | ------------- | ------------- | ------------- |
| max_len 2048 , attn_win 512 | Content Cell  | Content Cell  | Content Cell  |
| max_len 2048 , attn_win 1024 | Content Cell  | Content Cell  | Content Cell  |
| max_len 2048 , attn_win 2048 | Content Cell  | Content Cell  | Content Cell  |
| max_len 4096, attn_win 512 | Content Cell  | Content Cell  | Content Cell  |
| max_len 4096, attn_win 1024 | Content Cell  | Content Cell  | Content Cell  |
| max_len 4096, attn_win 2048 | Content Cell  | Content Cell  | Content Cell  |
| max_len 4096, attn_win 4096 | Content Cell  | Content Cell  | Content Cell  |
| max_len 16384, attn_win 512 | Content Cell  | Content Cell  | Content Cell  |
| max_len 16384, attn_win 1024 | Content Cell  | Content Cell  | Content Cell  |
| max_len 16384, attn_win 2048 | Content Cell  | Content Cell  | Content Cell  |
| max_len 16384, attn_win 4096 | Content Cell  | Content Cell  | Content Cell  |
| max_len 16384, attn_win 8192 | Content Cell  | Content Cell  | Content Cell  |
| max_len 16384, attn_win 16384 | Content Cell  | Content Cell  | Content Cell  |

**Comment by ydshieh** [comment]
For the record, here are the 2 scripts I used to measure running time (copied from yours with modification)

```python

import os
import json


def run(attention_window, steps, batch_size, max_length):

    os.system("rm -rf output.txt")
    os.system(f"python debug.py {attention_window} {steps} {batch_size} {max_length} > output.txt 2>&1")

    with open("output.txt") as fp:
        for line in fp:
            if f"{steps - 1}/{steps}" in line:
                line = line.strip()
                idx = line.find(f"{steps - 1}/{steps}")
                line = line[idx:]
                if "Initializing global" in line:
                    idx = line.find("Initializing global")
                    line = line[:idx]
                    line = line.strip()
                return line

res = {}
steps = 10

for batch_size in [1]:
    for max_length in [2048, 4096, 16384]:
        for attention_window in [512, 1024, 2048, 4096, 8192, 16384]:
            if attention_window > max_length:
                continue
            r = run(attention_window=attention_window, steps=steps, batch_size=batch_size, max_length=max_length)
            print(f"(attn_win: {attention_window}, batch_size: {batch_size}, max_len: {max_length}) --> {r}")
            print("=" * 40)

            res[f"(attn_win: {attention_window}, batch_size: {batch_size}, max_len: {max_length})"] = r

            with open("results.json", "w") as fp:
                json.dump(res, fp, indent=4, ensure_ascii=False)
```


```python
import sys

import torch
import datasets
import transformers
from transformers import BigBirdForSequenceClassification, Trainer, TrainingArguments, AutoTokenizer, AutoModel
from transformers.models.longformer.modeling_longformer import LongformerForSequenceClassification, LongformerConfig
from sklearn.metrics import accuracy_score
from torch.utils.data import Dataset

import logging
# logging.disable(logging.INFO)


def measure(attention_window, steps, batch_size, max_length):

    SLIDE_WIN_SIZE = attention_window

    STEPS = steps
    BATCH_SIZE = batch_size
    GRAD_ACCUMULATION_STEPS = 1
    LEN = max_length

    MODEL = 'allenai/longformer-base-4096'
    LONGFORMER = True
    CACHE_ROOT = "./"

    train_data, test_data = datasets.load_dataset('imdb', split=['train', 'test'], cache_dir=f'{CACHE_ROOT}/data')

    config = LongformerConfig.from_pretrained(MODEL, num_labels=2, return_dict=True)

    config.num_hidden_layers = 12
    config.hidden_size = 256
    config.num_attention_heads = 1
    config.intermediate_size = 1024

    config.attention_window = SLIDE_WIN_SIZE

    model = LongformerForSequenceClassification(config=config)
    tokenizer = AutoTokenizer.from_pretrained(MODEL, max_length=LEN, cache_dir=f'{CACHE_ROOT}/data')

    print("DEFAULT - Sliding window width across layers", model.config.attention_window)
    model.config.attention_window = SLIDE_WIN_SIZE
    print("UPDATED - Sliding window width across layers", model.config.attention_window)

    def tokenization(batched_text):
        return tokenizer(batched_text['text'], padding = 'max_length', truncation=True, max_length = LEN)

    train_data = train_data.map(tokenization, batched = True, batch_size = len(train_data))
    test_data = test_data.map(tokenization, batched = True, batch_size = len(test_data))

    train_data.set_format('torch', columns=['input_ids', 'attention_mask', 'label'])
    test_data.set_format('torch', columns=['input_ids', 'attention_mask', 'label'])

    def compute_metrics(pred):
        labels = pred.label_ids
        preds = pred.predictions.argmax(-1)
        acc = accuracy_score(labels, preds)
        return {'accuracy': acc}

    training_args = TrainingArguments(
        output_dir=f'{CACHE_ROOT}/results',
        # num_train_epochs=1,
        per_device_train_batch_size=BATCH_SIZE,
        max_steps=STEPS,
        gradient_accumulation_steps=GRAD_ACCUMULATION_STEPS,
        warmup_steps=160,
        weight_decay=0.01,
        learning_rate=2e-5,
        fp16=False,  # True,
        dataloader_num_workers=2,
        logging_strategy="steps",
        logging_steps=1,
    )

    trainer = Trainer(model=model, args=training_args, compute_metrics=compute_metrics, train_dataset=train_data)
    trainer.train()


if __name__ == "__main__":

    data = sys.argv[1:]
    print(data)
    data = [int(x) for x in data]
    measure(*data)
```

**Comment by omarwbassam** [comment]
No good

**Comment by krstp** [comment]
Great post. Thanks for sharing all the insights. Would you be able to share what was you input dataset size you were using, in particular number of rows and size of the text (tokens/words) you were using?

I was trying something similar on ~1mln input with large number of labels, initially with 4096 window, and 15 hours do not suffice to even train a single epoch.

**Comment by kiranvholla** [comment]
> Great post. Thanks for sharing all the insights. Would you be able to share what was you input dataset size you were using, in particular number of rows and size of the text (tokens/words) you were using?
> 
> I was trying something similar on ~1mln input with large number of labels, initially with 4096 window, and 15 hours do not suffice to even train a single epoch.

Apologies for the delayed response. It is almost 2 year old thread. I believe the dataset was for the Kaggle AI for code challenge here - https://www.kaggle.com/competitions/AI4Code. You can look at my writeup for details: https://www.kaggle.com/competitions/AI4Code/discussion/367140
